### PR TITLE
Festival Sect is back in 2022!

### PR DESCRIPTION
**Pull request by tralezab**
## Hey I've seen this before!

This is a revival of #61629, without power chord and with lullaby massively simplified. This removes the coding headache I had when making the pr originally and lets it be merged... as is!

## About The Pull Request

This was such a wonderful sect idea way back when and it never got finished which made me big sad. Ironically MOST of the work was already done, so I wanted to revive this

- [x] music barticles while playing sect songs?

## Why It's Good For The Game

Sparring sect was too much fun i'm turning this feature around! No but really, I got some requests from some players that there wasn't really a good "healing sect" and while I think healing is a bit too niche to base an entire sect and deity personality around something fun like a music theme is much more fitting while still filling that hole

image

## Changelog

:cl:
expansion: New Festival God sect! Sooth or serrate your friends and foes with the beat.
/:cl:

### CODE REVIEW

**Comment by Pickle-Coding** [comment]
> 

I just see a box.

**Comment by Shadowflame909** [comment]
Chaplain Bard!

**Comment by jughu** [comment]
I seduce the ash drake with my song

**Comment by Gboster-0** [comment]
> I seduce the ash drake with my song

its D&D bards all over again

**Review by Kylerace — CHANGES_REQUESTED** [review]

File: code/modules/religion/festival/instrument_rites.dm [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+/datum/religion_rites/song_tuner/sooth/song_effect(atom/song_player, datum/song/song_datum)
+	if(!song_datum)
+		return
+	for(var/mob/living/listener in song_datum.hearing_mobs)
+		if(listener.anti_magic_check(magic = FALSE, holy = TRUE))
+			continue
+		var/healy_juice = 0.6
+		if(listener.mind?.holy_role)
+			healy_juice*=3
+		listener.adjustBruteLoss(-healy_juice)
```
Comment (optimumtact) [inline code/modules/religion/festival/instrument_rites.dm:88]: is healing the only thing your brain is capable of coming up with as an effect?

we have an ENTIRE department for this.

File: code/modules/religion/festival/instrument_rites.dm [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+			continue
+		if(prob(20))
+			to_chat(listener, span_warning(pick("The music is putting you to sleep...", "The music makes you nod off for a moment.", "You try to focus on staying awake through the song.")))
+			listener.emote("yawn")
+		listener.blur_eyes(2)
+
+/datum/religion_rites/song_tuner/lullaby/finish_effect(atom/song_player, datum/song/song_datum)
+	for(var/mob/living/carbon/human/listener in song_datum.hearing_mobs)
+		to_chat(listener, span_danger("Wow, the ending of that song was... pretty..."))
+		listener.AdjustSleeping(5 SECONDS)
```
Comment (ATH1909) [inline code/modules/religion/festival/instrument_rites.dm:155]: vine boom sound effect

Comment (ATH1909) [reply code/modules/religion/festival/instrument_rites.dm:155]: should probably add an ear protection check to this just to make sure that you can't get a free gank on a seccie from it by cooking the song off screen for ~20 seconds

Comment (tralezab) [reply code/modules/religion/festival/instrument_rites.dm:155]: Is hearing is part of hearingmobs, will add antimagic

**Review by Twaticus — APPROVED** [review]

**Comment by TheBonded** [comment]
Can you explain what the sect does? Are you able to play music to buff nearby (non cult) people? And is there a matching null rod electric guitar axe to match the sect?

File: code/datums/components/smooth_tunes.dm [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	src.linked_songtuner_rite = linked_songtuner_rite
+	src.allow_repeats = allow_repeats
+	src.particles_path = particles_path
+	src.glow_color = glow_color
+
+/datum/component/smooth_tunes/Destroy(force, silent)
+	if(particle_holder)
+		QDEL_NULL(particle_holder)
+	qdel(linked_songtuner_rite)
+	. = ..()
```
Comment (MrMelbert) [inline code/datums/components/smooth_tunes.dm:34]:
```suggestion
	return ..()
```

File: code/datums/status_effects/buffs.dm [hunk 4]
```diff
@@ -348,9 +348,19 @@
 
 /datum/status_effect/antimagic
 	id = "antimagic"
+	status_type = STATUS_EFFECT_REFRESH
 	duration = 10 SECONDS
 	examine_text = "<span class='notice'>They seem to be covered in a dull, grey aura.</span>"
 
+/datum/status_effect/antimagic/on_creation(mob/living/new_owner, duration = 10 SECONDS)
+	src.duration = duration
+	. = ..()
```
Comment (MrMelbert) [inline code/datums/status_effects/buffs.dm:357]:
```suggestion
	return ..()
```

File: code/game/objects/effects/particle_holder.dm [hunk 5]
```diff
@@ -0,0 +1,65 @@
+///objects can only have one particle on them at a time, so we use these abstract effects to hold and display the effects. You know, so multiple particle effects can exist at once.
+///also because some objects do not display particles due to how their visuals are built
+/obj/effect/abstract/particle_holder
+	anchored = TRUE
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	layer = ABOVE_ALL_MOB_LAYER
+	///typepath of the last location we're in, if it's different when moved then we need to update vis contents
+	var/last_attached_location_type
+	var/datum/weakref/weak_attached
```
Comment (MrMelbert) [inline code/game/objects/effects/particle_holder.dm:9]: you auto-doc'd the other two but not this one

File: code/game/objects/effects/particle_holder.dm [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	var/atom/movable/additional_attached
+	if(weak_additional)
+		additional_attached = weak_additional.resolve()
+	if(attached)
+		attached.vis_contents -= src
+		UnregisterSignal(loc, list(COMSIG_MOVABLE_MOVED, COMSIG_PARENT_QDELETING))
+	if(additional_attached)
+		additional_attached.vis_contents -= src
+	QDEL_NULL(particles)
+	. = ..()
```
Comment (MrMelbert) [inline code/game/objects/effects/particle_holder.dm:36]:
```suggestion
	return ..()
```

File: code/modules/instruments/songs/_song.dm [hunk 7]
```diff
@@ -211,7 +211,8 @@
 	//we can not afford to runtime, since we are going to be doing sound channel reservations and if we runtime it means we have a channel allocation leak.
 	//wrap the rest of the stuff to ensure stop_playing() is called.
 	do_hearcheck()
-	SEND_SIGNAL(parent, COMSIG_SONG_START)
+	SEND_SIGNAL(parent, COMSIG_SONG_START, src)
+	SEND_SIGNAL(user, COMSIG_SONG_START, src)
```
Comment (MrMelbert) [inline code/modules/instruments/songs/_song.dm:215]: Surely these shouldn't be the same signal? `COMSIG_INSTRUMENT_SONG_START` and `COMSIG_MOB_START_PLAYING_SONG` or something

File: code/modules/instruments/songs/_song.dm [hunk 8]
```diff
@@ -222,15 +223,19 @@
...
+/datum/song/proc/stop_playing(finished = FALSE)
 	if(!playing)
 		return
 	playing = FALSE
 	if(!debug_mode)
 		compiled_chords = null
 	STOP_PROCESSING(SSinstruments, src)
-	SEND_SIGNAL(parent, COMSIG_SONG_END)
+	SEND_SIGNAL(parent, COMSIG_SONG_END, finished)
+	SEND_SIGNAL(music_player, COMSIG_SONG_END, finished)
```
Comment (MrMelbert) [inline code/modules/instruments/songs/_song.dm:238]: Same here - `COMSIG_INSTRUMENT_SONG_END` and `COMSIG_MOB_STOP_PLAYING_SONG`?

File: code/modules/instruments/songs/_song.dm [hunk 9]
```diff
@@ -239,23 +244,28 @@
...
+	current_chord++
+	if(current_chord <= length(compiled_chords))
+		return
+	if(!repeat)
+		stop_playing(TRUE)
+		return
+	repeat--
+	current_chord = 1
+	SEND_SIGNAL(parent, COMSIG_SONG_REPEAT, TRUE)
+	SEND_SIGNAL(music_player, COMSIG_SONG_REPEAT, TRUE)
```
Comment (MrMelbert) [inline code/modules/instruments/songs/_song.dm:268]: Aaaand same here?

File: code/modules/religion/festival/instrument_rites.dm [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+	song_start_message = span_warning("This music's making you feel drowsy...")
+	favor_cost = 40 //actually really strong
+	glow_color = "#83F6FF"
+	repeats_okay = FALSE
+	///assoc list of weakrefs to who heard the song and
+	var/list/listener_counter = list()
+
+/datum/religion_rites/song_tuner/lullaby/Destroy()
+	QDEL_NULL(listener_counter)
+	. = ..()
```
Comment (MrMelbert) [inline code/modules/religion/festival/instrument_rites.dm:128]:
```suggestion
	return ..()
```

File: code/modules/religion/festival/instrument_rites.dm [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+	song_invocation_message = "You've prepared a sleepy song!"
+	song_start_message = span_warning("This music's making you feel drowsy...")
+	favor_cost = 40 //actually really strong
+	glow_color = "#83F6FF"
+	repeats_okay = FALSE
+	///assoc list of weakrefs to who heard the song and
+	var/list/listener_counter = list()
+
+/datum/religion_rites/song_tuner/lullaby/Destroy()
+	QDEL_NULL(listener_counter)
```
Comment (MrMelbert) [inline code/modules/religion/festival/instrument_rites.dm:127]: `listener_count` is a list and you're just qdeling it directly, it should probably be a `Cut()` or something
```suggestion
	listener_counter.Cut()
```

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Comment by davethwave** [comment]
Welp, back to stale.

File: code/datums/status_effects/buffs.dm [hunk 12]
```diff
@@ -348,9 +348,19 @@
...
+	status_type = STATUS_EFFECT_REFRESH
 	duration = 10 SECONDS
 	examine_text = "<span class='notice'>They seem to be covered in a dull, grey aura.</span>"
 
+/datum/status_effect/antimagic/on_creation(mob/living/new_owner, duration = 10 SECONDS)
+	src.duration = duration
+	return ..()
+
+/datum/status_effect/antimagic/refresh(effect, duration = 10 SECONDS)
+	if(duration == -1)
```
Comment (Kylerace) [inline code/datums/status_effects/buffs.dm:360]: why exactly -1?

**Review by Kylerace — CHANGES_REQUESTED** [review]

File: code/modules/religion/festival/instrument_rites.dm [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+/datum/religion_rites/song_tuner/lullaby
+	name = "Spiritual Lullaby"
+	desc = "Sing a lullaby, tiring those around you, making them slower. At the end of the song, you'll put people who are tired enough to sleep."
+	particles_path = /particles/musical_notes/sleepy
+	song_invocation_message = "You've prepared a sleepy song!"
+	song_start_message = span_warning("This music's making you feel drowsy...")
+	favor_cost = 40 //actually really strong
+	glow_color = "#83F6FF"
+	repeats_okay = FALSE
+	///assoc list of weakrefs to who heard the song and
```
[the remark this answers is not in the document]
Comment (tralezab) [reply code/modules/religion/festival/instrument_rites.dm:123]: oh, that's easy, weakrefs and

File: code/datums/status_effects/buffs.dm (hunk 12 above)
Comment (tralezab) [reply code/datums/status_effects/buffs.dm:360]: -1 is how status effects set themselves to never expire

File: code/modules/religion/festival/instrument_rites.dm (hunk 13 above)
[the remark this answers is not in the document]
Comment (tralezab) [reply code/modules/religion/festival/instrument_rites.dm:123]: (will fix)

File: code/__HELPERS/_lists.dm [hunk 14]
```diff
@@ -777,6 +777,20 @@
 			: null\
 	)
 
+///takes a list, and returns a list with each of the contents as their weakref. Throws a stack trace if you feed it in non-instances
+/proc/weakref_list(list/real_list)
```
[the remark this answers is not in the document]
Comment (tralezab) [reply code/__HELPERS/_lists.dm:781]: Good point, lemmie blast it

**Comment by tralezab** [comment]
You don't have the right, O you don't have the right

**Comment by jughu** [comment]
try finger buthole

**Review by Kylerace — APPROVED** [review]
code good

**Review by MrMelbert — APPROVED** [review]
it's time

**Comment by Bawhoppen** [comment]
Woah it was finally mergeged

**Comment by thebestcodertoevercode** [comment]
Can we please get more variety? Buff / Heal songs would be amazing.

**Comment by LemonInTheDark** [comment]
stop necroing

**Pull request merged**